File: testdata/behavior/check.py
"""Shared response contract checks against generated Python clients."""
import json
import sys
from pathlib import Path
from typing import Any, cast

sys.path.insert(0, str(Path("gen").resolve()))
from typesafe_sdk import TypeSafeClient
from jev.ai.contract.v1.service_jev import ContractServiceClient
from jev.ai.contract.v1.contract_pb import EvaluateRequest, Container
from jev.ai.shared.v1.shared_pb import ExternalRequest

class Fake:
    def __init__(self, response: dict[str, Any], check_state: bool = True):
        self.response = response
        self.check_state = check_state

    def system_one(self, **kwargs: Any) -> dict[str, Any]:
        if self.check_state:
            assert kwargs["state"] == {"inputText":"hello", "inputCount":"5"}
        return self.response

cases = json.loads(Path("testdata/behavior/responses.json").read_text())
for tc in cases:
    client = ContractServiceClient(client=cast(TypeSafeClient, Fake(tc["response"])))
    try:
        result = client.evaluate_detailed(EvaluateRequest(input_text="hello", input_count=5))
    except (ValueError, TypeError):
        assert tc.get("error"), tc["name"]
        continue
    assert not tc.get("error"), tc["name"]
    assert json.loads(result.value.to_json()) == tc["expected"], (tc["name"], result.value)
    assert result.value.has_field("flag") and result.value.has_field("rating")
    assert result.response["model"] == "test-model"
    assert result.response["usage"] == tc["response"]["usage"]

client = ContractServiceClient(client=cast(TypeSafeClient, Fake({"answers":{"accepted":{"type":"noul", "noul":0.8}}}, False)))
assert client.external(ExternalRequest(input_text="hello")).accepted
assert client.nested(Container.NestedRequest(input_text="hello")).accepted
print(f"Python: {len(cases)} response fixtures and imported/nested RPCs passed")

File: testdata/behavior/check.ts
import assert from "node:assert/strict";
import fs from "node:fs";
import { create, toJson } from "@bufbuild/protobuf";
import { TypeSafeClient } from "@typesafe-ai/sdk";
import { ContractServiceClient } from "../../gen/jev/ai/contract/v1/service_jev.js";
import { EvaluateRequestSchema, EvaluateResponseSchema, Container_NestedRequestSchema } from "../../gen/jev/ai/contract/v1/contract_pb.js";
import { ExternalRequestSchema } from "../../gen/jev/ai/shared/v1/shared_pb.js";

const cases = JSON.parse(fs.readFileSync("testdata/behavior/responses.json", "utf8"));
async function main() {
  for (const tc of cases) {
    // Replace the network boundary only; use the real SDK types and generated Protobuf schemas.
    const sdk = new TypeSafeClient({apiKey:"test"});
    sdk.systemOne = (async (params: {state: unknown}) => {
      assert.deepEqual(params.state, {inputText:"hello", inputCount:"5"});
      return tc.response;
    }) as unknown as typeof sdk.systemOne;
    const client = new ContractServiceClient(sdk);
    const evaluate = () => client.evaluateDetailed(create(EvaluateRequestSchema, {inputText:"hello",inputCount:5n}));
    if (tc.error) { await assert.rejects(evaluate, tc.name); continue; }
    const result = await evaluate();
    assert.deepEqual(toJson(EvaluateResponseSchema,result.value),tc.expected,tc.name);
    assert.equal(typeof result.value.bigCount,"bigint");
    assert.equal(result.value.bigCount+1n,4n);
    assert.equal(result.response.model,"test-model");
    assert.deepEqual(result.response.usage,tc.response.usage);
  }
  const sdk = new TypeSafeClient({apiKey:"test"});
  sdk.systemOne = (async () => ({answers:{accepted:{type:"noul",noul:0.8}}})) as unknown as typeof sdk.systemOne;
  const client = new ContractServiceClient(sdk);
  assert.equal((await client.external(create(ExternalRequestSchema,{inputText:"hello"}))).accepted,true);
  assert.equal((await client.nested(create(Container_NestedRequestSchema,{inputText:"hello"}))).accepted,true);
  console.log(`TypeScript: ${cases.length} response fixtures and imported/nested RPCs passed`);
}
main().catch(e=>{console.error(e);process.exitCode=1;});
